File: src/resolver.edge.test.ts
import { test } from "node:test";
import { expect } from "expect";

import { createChecker } from "./checker.ts";
import { loadJdkStub } from "./jdkStub.ts";
import { getIdentifierAtPosition } from "./nodeAtPosition.ts";
import { createProgram } from "./program.ts";
import { resolveIdentifier } from "./resolver.ts";
import { type Identifier, type Symbol, SymbolFlags } from "./types.ts";

function ctxOf(text: string) {
  const program = createProgram();
  loadJdkStub(program);
  program.setOpenDocument("file:///T.java", text, 1);
  return { program, checker: createChecker(program), sf: program.getSourceFile("file:///T.java")! };
}

function resolveAt(text: string, needle: string, occurrence: number): Symbol | undefined {
  const { program, sf } = ctxOf(text);
  let offset = -1;
  for (let i = 0; i < occurrence; i++) offset = text.indexOf(needle, offset + 1);
  const id = getIdentifierAtPosition(sf, offset) as Identifier | undefined;
  return id ? resolveIdentifier(id, program) : undefined;
}

function resolveNameAt(text: string, needle: string, occurrence: number): Symbol | undefined {
  const ctx = ctxOf(text);
  let offset = -1;
  for (let i = 0; i < occurrence; i++) offset = text.indexOf(needle, offset + 1);
  const id = getIdentifierAtPosition(ctx.sf, offset) as Identifier | undefined;
  return id ? ctx.checker.resolveName(id) : undefined;
}

test("parameter shadows a field of the same name", () => {
  expect(resolveAt("class C { int x; int m(int x) { return x; } }", "x", 3)?.flags).toBe(
    SymbolFlags.Parameter,
  );
});

test("a local variable shadows a field", () => {
  expect(resolveAt("class C { int f; int m() { int f = 1; return f; } }", "f", 3)?.flags).toBe(
    SymbolFlags.LocalVariable,
  );
});

test("forward reference to a field declared later", () => {
  expect(resolveAt("class C { int m() { return later; } int later; }", "later", 1)?.flags).toBe(
    SymbolFlags.Field,
  );
});

test("field inherited through two levels of extends", () => {
  const sym = resolveAt(
    "class A extends B { int m() { return g; } }\nclass B extends Base {}\nclass Base { int g; }",
    "g",
    1,
  );
  expect(sym?.flags).toBe(SymbolFlags.Field);
});

test("default interface method is inherited by an implementor", () => {
  const sym = resolveNameAt(
    "interface I { default int d() { return 1; } }\nclass C implements I { void m() { d(); } }",
    "d",
    2,
  );
  expect(sym?.flags).toBe(SymbolFlags.Method);
});

test("enhanced-for variable resolves", () => {
  const sym = resolveAt(
    "class C { void m(java.util.List<String> xs) { for (String item : xs) { use(item); } } }",
    "item",
    2,
  );
  expect(sym?.flags).toBe(SymbolFlags.Parameter);
});

test("catch parameter resolves", () => {
  const sym = resolveAt(
    "class C { void m() { try {} catch (Exception ex) { use(ex); } } }",
    "ex",
    2,
  );
  expect(sym?.flags).toBe(SymbolFlags.Parameter);
});

test("typed lambda parameter resolves", () => {
  const sym = resolveAt("class C { Runnable r = (int arg) -> { use(arg); }; }", "arg", 2);
  expect(sym?.flags).toBe(SymbolFlags.Parameter);
});

test("enum constant resolves through member access", () => {
  const sym = resolveNameAt("enum E { A, B }\nclass C { E x = E.A; }", "A", 2);
  expect(sym?.flags).toBe(SymbolFlags.EnumConstant);
});

test("nested type resolves as a member type", () => {
  const sym = resolveAt("class O { Inner make() { return null; } class Inner {} }", "Inner", 1);
  expect(sym?.flags).toBe(SymbolFlags.Class);
});

test("overloaded method is one symbol with several declarations", () => {
  const sym = resolveAt(
    "class C { void m() { helper(); } void helper(){} void helper(int a){} }",
    "helper",
    1,
  );
  expect(sym?.declarations).toHaveLength(2);
});

test("method-scoped type parameter resolves", () => {
  const sym = resolveAt("class C { <T> T id(T x) { T y = x; return y; } }", "T", 4);
  expect(sym?.flags).toBe(SymbolFlags.TypeParameter);
});

File: togo/internal/compiler/resolver_edge_test.go
package compiler

import (
	"strings"
	"testing"
)

// Port of src/compiler/resolver.edge.test.ts. These cases load the JDK stub and
// use both the lexical resolver and the checker's member-aware resolveName.

func edgeCtx(text string) *checkerCtx {
	program := NewProgram()
	LoadJdkStub(program)
	program.SetOpenDocument("file:///T.java", text, 1)
	return &checkerCtx{program: program, checker: NewChecker(program), uri: "file:///T.java"}
}

func (ctx *checkerCtx) idAt(needle string, occ int) *Node {
	sf := ctx.program.GetSourceFile(ctx.uri)
	text := sf.AsSourceFile().Text
	offset := -1
	for i := 0; i < occ; i++ {
		offset = strings.Index(text[offset+1:], needle) + offset + 1
	}
	return GetIdentifierAtPosition(sf, offset)
}

// resolveAtEdge uses the plain lexical resolver (resolveIdentifier).
func resolveAtEdge(t *testing.T, text, needle string, occ int) *Symbol {
	t.Helper()
	ctx := edgeCtx(text)
	id := ctx.idAt(needle, occ)
	if id == nil {
		return nil
	}
	return ResolveIdentifier(id, ctx.program)
}

// resolveNameAtEdge uses the checker's member-aware resolveName.
func resolveNameAtEdge(t *testing.T, text, needle string, occ int) *Symbol {
	t.Helper()
	ctx := edgeCtx(text)
	id := ctx.idAt(needle, occ)
	if id == nil {
		return nil
	}
	return ctx.checker.ResolveName(id)
}

func TestParameterShadowsField(t *testing.T) {
	if sym := resolveAtEdge(t, "class C { int x; int m(int x) { return x; } }", "x", 3); sym == nil || sym.Flags != SymbolFlagsParameter {
		t.Errorf("got %v, want parameter", sym)
	}
}

func TestLocalShadowsFieldEdge(t *testing.T) {
	if sym := resolveAtEdge(t, "class C { int f; int m() { int f = 1; return f; } }", "f", 3); sym == nil || sym.Flags != SymbolFlagsLocalVariable {
		t.Errorf("got %v, want local", sym)
	}
}

func TestForwardReferenceField(t *testing.T) {
	if sym := resolveAtEdge(t, "class C { int m() { return later; } int later; }", "later", 1); sym == nil || sym.Flags != SymbolFlagsField {
		t.Errorf("got %v, want field", sym)
	}
}

func TestFieldInheritedTwoLevels(t *testing.T) {
	sym := resolveAtEdge(t, "class A extends B { int m() { return g; } }\nclass B extends Base {}\nclass Base { int g; }", "g", 1)
	if sym == nil || sym.Flags != SymbolFlagsField {
		t.Errorf("got %v, want field", sym)
	}
}

func TestDefaultInterfaceMethodInherited(t *testing.T) {
	sym := resolveNameAtEdge(t, "interface I { default int d() { return 1; } }\nclass C implements I { void m() { d(); } }", "d", 2)
	if sym == nil || sym.Flags != SymbolFlagsMethod {
		t.Errorf("got %v, want method", sym)
	}
}

func TestEnhancedForVariableResolves(t *testing.T) {
	sym := resolveAtEdge(t, "class C { void m(java.util.List<String> xs) { for (String item : xs) { use(item); } } }", "item", 2)
	if sym == nil || sym.Flags != SymbolFlagsParameter {
		t.Errorf("got %v, want parameter", sym)
	}
}

func TestCatchParameterResolves(t *testing.T) {
	sym := resolveAtEdge(t, "class C { void m() { try {} catch (Exception ex) { use(ex); } } }", "ex", 2)
	if sym == nil || sym.Flags != SymbolFlagsParameter {
		t.Errorf("got %v, want parameter", sym)
	}
}

func TestTypedLambdaParameterResolves(t *testing.T) {
	sym := resolveAtEdge(t, "class C { Runnable r = (int arg) -> { use(arg); }; }", "arg", 2)
	if sym == nil || sym.Flags != SymbolFlagsParameter {
		t.Errorf("got %v, want parameter", sym)
	}
}

func TestEnumConstantThroughMemberAccess(t *testing.T) {
	sym := resolveNameAtEdge(t, "enum E { A, B }\nclass C { E x = E.A; }", "A", 2)
	if sym == nil || sym.Flags != SymbolFlagsEnumConstant {
		t.Errorf("got %v, want enum constant", sym)
	}
}

func TestNestedTypeAsMemberType(t *testing.T) {
	sym := resolveAtEdge(t, "class O { Inner make() { return null; } class Inner {} }", "Inner", 1)
	if sym == nil || sym.Flags != SymbolFlagsClass {
		t.Errorf("got %v, want class", sym)
	}
}

func TestOverloadedMethodOneSymbol(t *testing.T) {
	sym := resolveAtEdge(t, "class C { void m() { helper(); } void helper(){} void helper(int a){} }", "helper", 1)
	if sym == nil || len(sym.Declarations) != 2 {
		t.Errorf("got %v declarations, want 2", sym)
	}
}

func TestMethodScopedTypeParameterResolves(t *testing.T) {
	sym := resolveAtEdge(t, "class C { <T> T id(T x) { T y = x; return y; } }", "T", 4)
	if sym == nil || sym.Flags != SymbolFlagsTypeParameter {
		t.Errorf("got %v, want type parameter", sym)
	}
}
